zerver/lib/actions.py

```diff
@@ -1364,8 +1364,9 @@ def do_send_messages(messages_maybe_none: Sequence[Optional[MutableMapping[str,
 
         # Add members of the mentioned user groups into `mentions_user_ids`.
         for group_id in message['message'].mentions_user_group_ids:
-            members = message['mention_data'].get_group_members(group_id)
-            message['message'].mentions_user_ids.update(members)
+            ids = message['mention_data'].get_group_members(group_id)
+            group_mentions = {id: group_id for id in ids if id not in message['message'].mentions_user_ids}
+            message['message'].mentions_user_ids.update(group_mentions)
```

Comment:
I'm not sure I like this overall strategy for how to send the data.  I think what we want is to send just the fact that the user was directly mentioned if that's the case, and for users who weren't, just the user group details for the group through which they were mentioned.  So I think what this means is we should be changing `mentions_user_ids` to be JUST the users who were directly mentioned, and adding a `user_group_mentions` structure mapping each mentioned user ID to the user group ID that's mentioning them.  This would be a cleaner API that wouldn't require the messy logic in the event queue side.

Reply:
@timabbott  Yes adding a user_group_mentions structure would definitely be be more clearer though i was trying to follow the method mentioned in [issue](https://github.com/zulip/zulip/issues/13080). Which says that the `benefit of this model is that we don't have to change most downstream code that just looks at keys in mentions_user_ids.`

Reply:
Yeah, having seen how this looks, I think my prediction of the best way to do this written at the time was wrong :)

Reply:
@timabbott do we also need to add group_mentioned flag in `AbstractUserMessage`?

Reply:
@timabbott Any updates?

Reply:
Yeah, I think it's worth the migration to that model.  I guess we'll need to be a bit thoughtful about how the mobile apps see it. 

@arpit551 do you have time to work on this again?  

@abhijeetbodas2001 FYI, since you're thinking about this code path, and might be a good reviewer for this project now that you have this part of the codebase swapped in.